app/src/main/res/values/strings.xml

```diff
@@ -260,7 +260,7 @@
   <!-- AddProfileActivity -->
   <string name="add_profile_activity_label">Add Profile</string>
   <string name="add_profile_title">Add Profile</string>
-  <string name="add_profile_input_name">Name*</string>
+  <string name="add_profile_input_name"><font color='@android:color/transparent'>*</font>Name*<font color='@android:color/transparent'>required</font></string>
```

Comment:
Won't this result in the label generally not being visible? I'm also a bit confused--wasn't there a different approach you were looking into earlier where we use a different content description that doesn't include the asterisk?

Reply:
- The label is visible, as shown in the video below.
   [samaste.webm](https://user-images.githubusercontent.com/31986629/191627085-eb971ed5-c0f3-4f96-983d-f1dd5c9372bd.webm)

- According to the issue(https://github.com/oppia/oppia-android/issues/4322), "Talkback should read required or mandatory as a label for mandatory fields instead of the asterisk". This is addressed by the current **simple** solution. 

- However, I've tried to add the hint in TextInputLayout and the hint in the TextInputEditText approach. The results shown in the video below work (does not pronounce asterisk), but require is not pronounced,  and the UI is altered in a way.
[samaste2.webm](https://user-images.githubusercontent.com/31986629/191627314-f2f9a075-5ee1-4f3f-9b33-210bfe22cc4d.webm)

- I've also tried the Content description approach; 
     - content description in the TextInputLayout and hint in the TextInputEditText does not work(mentions asterisk). And the content description is ignored.
     - But according to the android accessibility doc(https://support.google.com/accessibility/android/answer/6378120?hl=en), 
       "Defining an android:contentDescription on any EditText or editable TextView may interfere with an accessibility service's ability to describe, navigate, and interact with text that a user enters within the item."
     - Using contentDescription via methods described here will for sure work, but at the cost of compromising code quality through deleting TextInputLayout and using a textview and and ordinary edittext. More so, on the same xml page, there exists 2 other TextInputLayout that have and astrisks, therefore, the contentDescription approach is unnecessary.

Reply:
Thanks for the detailed write-up @Ryggs. I think the issue with the current approach is it's relying on visual mark-up (font-color) to influence an entirely unrelated system (a11y announcements). This feels like  a fragile solution that may not work consistently across different system & screenreader configurations, or potentially long-term.

The enumeration of additional approaches is really helpful.

At this point, I have to start wondering whether '*' is actually the correct representation to be using for a required field. The general advice when deciding on an visually impaired user's experience is to consider whether challenges they face may also be potential issues for those users who are not visually impaired. '*' while conventional is not actually a very clear indicator of required (even with the extra labeling), and we could possibly be more explicit in a way that benefits both types of users.

@mschanteltc what do you think about us removing the '* Required' field and switching the hint text for the profile name to be "Name (Required)"? Or, do you have any alternate suggestions that don't involve using symbols since screenreaders may not read them out correctly?

Reply:
Would it be possible to have [helper text](https://material.io/components/text-fields#usage) "Required" like in the image below?:
![image](https://user-images.githubusercontent.com/22177674/192079202-30723956-3792-4c52-924f-553319396560.png)

Reply:
That would be much clearer if the accessibility story works. Per https://material.io/components/text-fields/android#using-text-fields (see under "Adding a prefix/suffix to a text field") we can actually set either text underneath or suffix text:

![Screenshot showing different hint text options in Material](https://lh3.googleusercontent.com/vY-Jq5nFRhU8i4kevJlmI8e52vtR_KhTzPx6kQTpKIB7-XwXXcjGYMvFISLe9dGipSeh2JMTHsO3KSK0RAFyehw1XBNewpItJ_BUFvGNDB6_KDuCuR1Z=w1064-v0)

Which do you think we should use @mschanteltc?

Reply:
Also @Ryggs could you try out the above solution (for both prefix & suffix texts) and see what the Talkback behavior is like for those values? We need to make sure that they're read out properly.

Reply:
@BenHenning , 

1. Prefix: * is announced only in the first column.

| app:prefixText="*" | app:prefixText="*" && app:expandedHintEnabled="false"|
| ------------- | ------------- |
| * is anounced | * not announced | 
| Label appears in input are then up top when expanded | Label appears expanded, up to the Text area |
|[prefix.webm](https://user-images.githubusercontent.com/31986629/192989124-c8b32ab1-1dcc-4304-bed1-75ae7e590d7c.webm)  |  [expHintPrefix.webm](https://user-images.githubusercontent.com/31986629/192991466-59f75cf5-6ac9-4fb1-b745-f258082f6511.webm) |

2. Suffix:  * is announced only in the first column, 

|  app:suffixText="*" | app:suffixText="*" && app:expandedHintEnabled="false" |
| ----- | ----- |
| * is anounced | * not announced | 
| Label appears in input are then up top when expanded | Label appears expanded, up to the Text area |
| [sffix.webm](https://user-images.githubusercontent.com/31986629/192991191-33be6b6d-b275-4bb3-94d5-2c4f808e2e69.webm)| [suffixenable.webm](https://user-images.githubusercontent.com/31986629/192990340-2bf25d8f-bb91-4393-ad15-c0bab9063b68.webm)  |

3. Helper text: This doesn't seem to have any effect with talkback.

|  app:helperText="Required *" |  app:helperText="Required *" and app:helperTextEnabled="true" |
| ----- | ----- |
| [helper.webm](https://user-images.githubusercontent.com/31986629/192991664-db0e3ff9-e6c3-44fd-9bed-36a48424fe5e.webm)  | [helperenabled.webm](https://user-images.githubusercontent.com/31986629/192991792-113ee355-bcfd-42a7-adbf-772a9abd4e73.webm)  |

Reply:
Thanks @Ryggs.

I actually think we can omit the '*' entirely if we go with this approach since we can make it clear in-context what's required, but this context really helps with understanding our choices.

@mschanteltc what's your preference here between prefix & suffix labels (per the earlier resource I linked)?

Reply:
As discussed in chat today, @Ryggs I think you can go ahead with a suffix solution & then wrap up this PR while we wait for Chantel's feedback. Switching to prefix should be straightforward to do at the end if we need to.

Reply:
Sure @BenHenning

Reply:
Hey @BenHenning , according to your message above ~~ " The general advice when deciding on an visually impaired user's experience is to consider whether challenges they face may also be potential issues for those users who are not visually impaired. " I want to show you that the latter category(not visually impaired.) does not get affected.

According to the issue, talkback announces *. In the page UI flow, before all edit boxes, under the profile photo section as seen in the video above, we have a text field marked *Required. That's used to guide the user(sighted or not).

To sighted person, they can look at fields marked with * and know that it's required. This works as it should.
To a visually impaired person, talkback announces that text view(*Required) as ~~ "Fields marked with an * are required".  Which I also believe works as it should.

By replacing/silencing *, the latter group(visually impaired) will be misguided.

Reply:
@Ryggs I think there might be a miscommunication or misunderstanding here. I've left a comment in code, but I'll also mention it here: we don't need to use '*' at all anymore with suffix text being available, and given that, we should get rid of it as it'll improve both the screenreader and non-screenreader usability.

Reply:
The changes have been implemented.

Reply:
> That would be much clearer if the accessibility story works. Per https://material.io/components/text-fields/android#using-text-fields (see under "Adding a prefix/suffix to a text field") we can actually set either text underneath or suffix text:
> 
> ![Screenshot showing different hint text options in Material](https://camo.githubusercontent.com/faaefc77eeb2cd9d53a9e53339f50faf6a144e7563ee9d1f9a4de9baf93dad59/68747470733a2f2f6c68332e676f6f676c6575736572636f6e74656e742e636f6d2f76592d4a71356e465268553869346b65764a6c6d49386535327674525f4b68547a5078366b5154704b4942372d58775858636a47594d764649534c653964476970536568324a4d5448734f334b534b30524146796568773158424e65777049744a5f42554676474e4442365f4b4475437552315a3d77313036342d7630)
> 
> Which do you think we should use @mschanteltc?

Since width space is limited on mobile devices, let's use 'text underneath' the input field.

Reply:
Sounds good--thanks @mschanteltc!

@Ryggs does this unblock you?

Reply:
> > That would be much clearer if the accessibility story works. Per https://material.io/components/text-fields/android#using-text-fields (see under "Adding a prefix/suffix to a text field") we can actually set either text underneath or suffix text:
> > ![Screenshot showing different hint text options in Material](https://camo.githubusercontent.com/faaefc77eeb2cd9d53a9e53339f50faf6a144e7563ee9d1f9a4de9baf93dad59/68747470733a2f2f6c68332e676f6f676c6575736572636f6e74656e742e636f6d2f76592d4a71356e465268553869346b65764a6c6d49386535327674525f4b68547a5078366b5154704b4942372d58775858636a47594d764649534c653964476970536568324a4d5448734f334b534b30524146796568773158424e65777049744a5f42554676474e4442365f4b4475437552315a3d77313036342d7630)
> > Which do you think we should use @mschanteltc?
> 
> Since width space is limited on mobile devices, let's use 'text underneath' the input field.

Suggestion: We can hide the suffix, every time the user tries type to Or types something. This way, we can still cater to both categories of users(talkback/no-talkback ~~ prefix).  This is illustrated by the video below.

https://user-images.githubusercontent.com/31986629/195359552-28bcc61b-997d-47ee-89d1-30f72fad9a13.mp4

Reply:
However, I have already implemented the requested changes.
Removed suffix and used

https://user-images.githubusercontent.com/31986629/195363886-f36e593c-a1bf-4fcd-b60e-24c38f0d1eed.mp4

 underneath text(helper text)

Reply:
@Ryggs does your first video have Talkback text output enabled? It doesn't seem like it, and it's important to show that hiding doesn't change the readout behavior.

That being said, is the auto-hide something custom or something that TextInputLayout does automatically for us?

@mschanteltc do you have any thoughts on the auto-hide behavior shown above for the suffix approach? Or, should we still prioritize the prefix/underneath text?

Reply:
Hiding does not change read-out behavior, as shown by the video below.
The auto Hiding is custom, required is hidden by the snippet below(inside the onTextChanged listener lambda).

`binding.addProfileActivityUserNameEditText.onTextChanged { name ->
      name?.let {
        if (
          profileViewModel.nameErrorMsg.get()?.isNotEmpty()!! &&
          profileViewModel.inputName.get() == it
        ) {
          profileViewModel.requiredField.set("")
          ...
        } else {
          profileViewModel.requiredField.set("")
          ...
        }
      }
    }`

Talk-back output for the first video is enabled, however, the required is not pronounced. Below is the same video but much clear.

https://user-images.githubusercontent.com/31986629/195666246-c28f5dcb-807b-4fdb-bbb3-3c39439be1e0.mp4

While required is mentioned in the below video

https://user-images.githubusercontent.com/31986629/195666551-fa3de52a-f021-4ba9-9d8e-34e7f0fbdea2.mp4

Reply:
Hi @BenHenning I have pushed the helper text implementation. PTAL

Reply:
> @Ryggs does your first video have Talkback text output enabled? It doesn't seem like it, and it's important to show that hiding doesn't change the readout behavior.
> 
> That being said, is the auto-hide something custom or something that TextInputLayout does automatically for us?
> 
> @mschanteltc do you have any thoughts on the auto-hide behavior shown above for the suffix approach? Or, should we still prioritize the prefix/underneath text?

My concern for the suffix approach is that the suffix acts as a placeholder and disappears when their is input. Based on this [article](https://www.nngroup.com/articles/form-design-placeholders/), placeholders causes relevant information to 'disappear.' I would still prioritize the prefix because it provides users necessary information regardless if there is input or not.

Reply:
Thanks @mschanteltc, having said that what is the final count on the decision for this issue? Prefix or Suffix?

Reply:
Thanks @mschanteltc!

@Ryggs I think prefix is the approach we want to take. I think both Chantel and I provided reasons for it, and it doesn't seem like the suffix approach works as well. Let's go with prefix.

Reply:
Hi @BenHenning , I've implemented the prefix approach. PTAL

Reply:
So I think this may be a case where I didn't understand what was meant by prefix. The latest videos are not showing what @mschanteltc suggested above--the required text needs to go underneath the TextView, not within it. Given that this came from material design (see my link toward the top of this thread), it should be possible to set text below the input box not within it.

Is the issue the helper text not being announced? As far as I can tell from TextInputLayout's source, it should be setting its helper text as polite which means Talkback will generally announce it (but may skip it under some circumstances). That means it should be announced here.

Reply:
The prefix has since been removed and replaced with the text helper approach. PTAL

Reply:
Thanks @Ryggs. Three things:
- Can you update your non-Talkback after-change video to include trying to submit no entry? It should trigger an error, and it'd be good to see how that looks with the change.
- Can you recapture your after-change Talkback video to use side-swiping for navigation instead of direct tapping? I don't expect it to change anything, but I want to make sure. Talkback users won't be directly tapping on specific elements.
- Can you try updating to the latest material core component library (whichever contains the TextInputLayout)? It seems wrong that the announcement isn't happening even if it is polite. We may need to check this in an file a bug against the Material team, but let's see if we can figure out how to get it to be hooked up correctly first.

Reply:
Regarding the first point. According to the code, the button is only enabled when there is text in the input name editbox. Therefore it will not trigger any error on any editbox or required error messages. However, when you type in the name editbox, and(or) mismatch the pin passwords, then an error is thrown. As show in the video above.

The second point, swipe navigation works. More so, the required field is pronounced when the user swipes, therefore announcement is happening. (i guess this answers the third point). 

@BenHenning PTAL

Reply:
Thanks @Ryggs. Can you update the two "with change" videos in the PR description to show both of these? I think having that frame of reference at the top level of the PR is actually important to make it easier to see for other reviews, or other people who might look at this PR in the future (after its merged). Better to not bury important details within conversation threads.

Reply:
The videos showcasing the two points have been added. PTAL